### PR TITLE
Fix limit memory usage of `XML::Error.errors`

### LINKED ISSUE #14934: `XML::Error.errors` leaks memory
`XML::Error` has a class variable `@@errors` which accumulates error messages from calls into `libxml2`.

It's accessible via the class method `XML::Error.errors` which returns the current items and resets the instance variable.
This global error handling is a legacy mechanism. Its use is deprecated. It was replaced by a more localized variant in #12663.
Still, we're keeping the old API around for backwards compatibility.

But when you don't use the deprecated API - as recommended - that means `.errors` never resets `@@errors` and the array with error messages only grows and wastes memory.

Example:
```cr
require "xml"

class XML::Error
  XML::Reader.new("</foo>").read
  @@errors # => [#<XML::Error:StartTag: invalid element name>]

  XML::Reader.new("<bar").read
  @@errors # => [#<XML::Error:StartTag: invalid element name>, #<XML::Error:Couldn't find end of Start Tag bar>]
end
```

This bug is a bit of a dilemma: We keep supporting an old API but not using it causes it to malfunction.

Maybe we could try to come up with some clever mechanism to clear the errors array every now and then or cap its size to keep extend in check. But I don't think that's going to work sufficiently well. For users of the new API it would still be wasteful. And we cannot guarantee stability for users of the old API.
So IMO our only option is to cut off the global error collection.
We could keep the deprecated method `XML::Error.errors` and have it always return `nil`. But at that point it has no meaningful function anymore and we could just drop it entirely. That'll at least mean a hard break for any code that still uses the deprecated API.
If necessary, it could be possible to provide the previous behaviour as an opt-in.

### PR DESCRIPTION
**Pull request by straight-shoota**
This patch limits the size of `XML::Error.errors` to contain only the last 10 error messages.
A new class property `XML::Error.max_error_capacity` allows configuration of the error capacity. This property is already marked as deprecated because it'll vanish together with `XML::Error.errors`.

Resolves #14934
Closes https://github.com/crystal-lang/crystal/pull/14936

### CODE REVIEW

File: src/xml/error.cr [hunk 1]
```diff
@@ -11,19 +11,27 @@ class XML::Error < Exception
     super(message)
   end
 
-  @@errors = [] of self
+  @[Deprecated("This class property is deprecated. XML errors are accessible directly in the respective context via `XML::Reader#errors` and `XML::Node#errors`.")]
```
Comment (beta-ziliani) [inline src/xml/error.cr:20]: This must be split into the accessors, and deprecate those.

**Review by beta-ziliani — CHANGES_REQUESTED** [review]

**Review by beta-ziliani — APPROVED** [review]

File: src/xml/error.cr [hunk 2]
```diff
@@ -11,19 +11,36 @@ class XML::Error < Exception
     super(message)
   end
 
-  @@errors = [] of self
+  @@max_error_capacity = 5
+
+  @[Deprecated("This class property is deprecated. XML errors are accessible directly in the respective context via `XML::Reader#errors` and `XML::Node#errors`.")]
+  def self.max_error_capacity=(@@max_error_capacity)
+  end
```
Comment (ysbaddaden) [inline src/xml/error.cr:18]: **Question**: if I want to disable the behavior, which is the way forward, I have to set it this value to 0, but then I get a deprecation notice :confused:

File: src/xml/error.cr [hunk 3]
```diff
@@ -11,19 +11,36 @@ class XML::Error < Exception
...
+
+  @[Deprecated("This class property is deprecated. XML errors are accessible directly in the respective context via `XML::Reader#errors` and `XML::Node#errors`.")]
+  def self.max_error_capacity
+    @@max_error_capacity
+  end
+
+  @@errors = Deque(self).new(max_error_capacity)
```
Comment (ysbaddaden) [inline src/xml/error.cr:25]: **Question**: this initializer will run before user code, won't `@max_error_capacity` always be the default 5, whatever I set?

**Comment by ysbaddaden** [comment]
This is limiting the leak (it won't keep growing) but it' still allocating pointless strings that will have to be collected... for how long are we gonna keep this?

I'm leaning more to have `XML::Error.errors` always return `nil` from now on, with the documentation stating that it used to return errors but doesn't since v1.14, or just plain drop the method.

File: src/xml/error.cr (hunk 2 above)
Comment (straight-shoota) [reply src/xml/error.cr:18]: Hm, I don't think you should actually need to disable it explicitly.
You should of course be able to do that if you want. And in that case this method should not raise a deprecation message. But we want to get rid of this method as well, so it should still be deprecated.
Maybe we should mark it as `Experimental` instead? 🤔
Or don't annotate it at all but document it that it should always be guarded by `compare_version(Crystal::VERSION, "1.13.2") > 0) && compare_version(Crystal::VERSION, "2.0.0") < 0)`.

File: src/xml/error.cr (hunk 3 above)
Comment (straight-shoota) [reply src/xml/error.cr:25]: Argh, yeah it will 🤦

**Comment by straight-shoota** [comment]
Yeah, I'm all in for disabling the feature entirely. This patch is just adding more complexity to something we need to get rid of anyway.
https://github.com/crystal-lang/crystal/pull/14936 is ready as an alternative.

**Comment by bararchy** [comment]
Should this be closed in favor of https://github.com/crystal-lang/crystal/pull/14936 ?

**Pull request closed without merging**